### PR TITLE
Use common code to set expected shard size

### PR DESCRIPTION
**Pull request by idegtiarenko**
This change uses common component to set expected shard size when initializing shard from GatewayAllocator

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/gateway/BaseGatewayShardAllocator.java [hunk 1]
```diff
@@ -58,26 +59,14 @@ public void allocateUnassigned(
...
             );
         } else {
             unassignedAllocationHandler.removeAndIgnore(allocateUnassignedDecision.getAllocationStatus(), allocation.changes());
         }
     }
 
-    protected static long getExpectedShardSize(ShardRouting shardRouting, RoutingAllocation allocation) {
-        if (shardRouting.primary()) {
-            if (shardRouting.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
-                return allocation.snapshotShardSizeInfo().getShardSize(shardRouting, ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE);
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/gateway/BaseGatewayShardAllocator.java:72]: I believe we could recover replicas from snapshots as well, do we?

File: server/src/main/java/org/elasticsearch/gateway/BaseGatewayShardAllocator.java [hunk 2]
```diff
@@ -58,26 +59,14 @@ public void allocateUnassigned(
...
-            } else {
-                return ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE;
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/gateway/BaseGatewayShardAllocator.java:74]: In future we could estimate new shard size from other sources as well, for example shard size forecast. I am going to open separate pr for that

File: server/src/main/java/org/elasticsearch/gateway/BaseGatewayShardAllocator.java (hunk 1 above)
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/gateway/BaseGatewayShardAllocator.java:72]: Above comment is actually not correct :)

File: server/src/main/java/org/elasticsearch/cluster/routing/ExpectedShardSizeEstimator.java [hunk 3]
```diff
@@ -48,7 +48,7 @@ public static long getExpectedShardSize(
             && shard.active() == false
             && shard.recoverySource().getType() == RecoverySource.Type.LOCAL_SHARDS) {
             return getExpectedSizeOfResizedShard(shard, defaultValue, indexMetadata, clusterInfo, metadata, routingTable);
-        } else if (shard.unassigned() && shard.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
+        } else if (shard.primary() && shard.active() == false && shard.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
             return snapshotShardSizeInfo.getShardSize(shard, defaultValue);
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/ExpectedShardSizeEstimator.java:52]: This is changed to align with code in `SnapshotShardSizeInfo#getShardSize`
https://github.com/elastic/elasticsearch/blob/e1c284a7c3aac0a238e8ac5fa4fe8a29f33e5685/server/src/main/java/org/elasticsearch/snapshots/SnapshotShardSizeInfo.java#L27-L29

**Review by fcofdez — APPROVED** [review]

**Review by DaveCTurner — APPROVED** [review]
LGTM2

**Pull request merged**